### PR TITLE
Feature/codeform

### PR DESCRIPTION
**Pull request by ramirobg94**

### CODE REVIEW

File: src/form/Attributes/Code/Preview/CodeCheckbox/index.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        event.preventDefault();
+
+        let _node = node;
+
+        if (NodeDefs.isMultiple(nodeDef) && Objects.isEmpty(_node)) {
+          _node = nodes.find(({value}) => Objects.isEmpty(value));
+        }
+        if (!NodeDefs.isMultiple(nodeDef) && Objects.isEmpty(_node)) {
+          _node = node || nodes[0];
+        }
```
Comment (SteRiccio) [inline src/form/Attributes/Code/Preview/CodeCheckbox/index.js:84]: this code could be extracted to a function like "determineNode" or something like that

File: src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import {call} from 'redux-saga/effects';
+
+import handleCreateNodeAndDescendants from '../createNodeAndDescendants';
+import handleUpdateNode from '../updateNode';
+
+function* handleCreateNodeAndDescendantsWithValue({payload} = {}) {
```
Comment (SteRiccio) [inline src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js:6]: createNodeAndDescendants makes sense when the node is an entity. This function should be called just something like "createAttributeWithValue"

File: src/state/nodes/sagas/removeNode/index.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  const [record, recordNodesByUuid] = yield all([
+    select(formSelectors.getRecord),
+    select(formSelectors.getRecordNodesByUuid),
+  ]);
+  const recordWithNodes = {...record, nodes: {...(recordNodesByUuid || {})}};
+
+  const recordUpdated = Records.removeNode(node)(recordWithNodes);
+
+  let nodesToDelete = [];
+  const newRecordsUuids = Object.keys(recordUpdated.nodes);
```
Comment (SteRiccio) [inline src/state/nodes/sagas/removeNode/index.js]: newRecordsUuids is actually a newRecordNodesUuids

File: src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js (hunk 2 above)
Comment (ramirobg94) [reply src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js:6]: at the end we are creating a node, and this function is used also with entities

File: src/state/nodes/sagas/removeNode/index.js (hunk 3 above)
Comment (ramirobg94) [reply src/state/nodes/sagas/removeNode/index.js]: true

File: src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  const nodes = yield call(handleCreateNodeAndDescendants, {
+    payload: {
+      nodeDef,
+      parentNode,
+    },
+  });
+
+  const updatedNodes = Object.values(nodes);
+
+  const updatedNodeWithValue = Object.assign({}, updatedNodes[0], {value});
```
Comment (SteRiccio) [inline src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js:18]: assign the value only if specified (not necessary when creating an entity) or check if the node def is an attribute before doing this

File: src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js (hunk 2 above)
Comment (SteRiccio) [reply src/state/nodes/sagas/createNodeAndDescendantsWithValue/index.js:6]: then call the function handleCreateNodeAndDescendants, or handleCreateNode... "with value" as I was saying applies only to attributes

File: src/state/nodes/sagas/removeNode/index.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  const recordUpdated = Records.removeNode(node)(recordWithNodes);
+
+  let nodesToDelete = [];
+  const recordUpdatedNodes = Object.keys(recordUpdated.nodes);
+  Object.keys(recordNodesByUuid).map(nodeUuid => {
+    if (!recordUpdatedNodes.includes(nodeUuid)) {
+      nodesToDelete.push(recordNodesByUuid[nodeUuid]);
+    }
+  });
+
```
Comment (ramirobg94) [inline src/state/nodes/sagas/removeNode/index.js:29]:
```suggestion
// this code is needed if you dont have the index as arena-core search descendants by index
const currentDescentands = yield select(state =>
    formSelectors.getNodeDescendants(state, node),
  );

  nodesToDelete = nodesToDelete.concat(currentDescentands);
```

File: src/state/nodes/sagas/removeNode/index.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  const recordUpdated = Records.removeNode(node)(recordWithNodes);
+
+  let nodesToDelete = [];
+  const recordUpdatedNodes = Object.keys(recordUpdated.nodes);
+  Object.keys(recordNodesByUuid).map(nodeUuid => {
+    if (!recordUpdatedNodes.includes(nodeUuid)) {
+      nodesToDelete.push(recordNodesByUuid[nodeUuid]);
+    }
+  });
```
Comment (ramirobg94) [inline src/state/nodes/sagas/removeNode/index.js]:
```suggestion
  Object.keys(recordNodesByUuid).forEach(nodeUuid => {
    if (!recordUpdatedNodes.includes(nodeUuid)) {
      nodesToDelete.push(recordNodesByUuid[nodeUuid]);
    }
  });
```

**Review by SteRiccio — APPROVED** [review]

**Pull request merged**